// new-scratch3-api/lib/api/index.js
"use strict";

const {
  getJSON
} = require("../request");

class API {
  get conference() {
    return new Conference();
  }

  get users() {
    return new Users();
  }

  get site() {
      return new Site();
  }

  get proxy() {
      return new Proxy();
  }
}

class Conference {
  async scheduleForDay(day, zidx = true) {
    let days = ["Saturday", "Sunday", "Monday", "Tuesday", "Wednesday", "Thursday", "Friday"].map(function (day) {
      return day.toLowerCase();
    });

    let d = function () {
      switch (typeof day) {
        case "string":
          if (days.includes(day)) {
            return day;
          } else {
            throw new Error("Invalid day of week");
          }

        case "number":
          if (Math.floor(day) !== day) throw new Error("Day of week must be an integer");

          if (zidx) {
            if (day < 6 && day >= 0) {
              return days[day];
            }

            throw new Error(`${day} is not a valid day of the week`);
          } else {
            if (day < 7 && day >= 1) {
              return days[day - 1];
            }

            throw new Error(`${day} is not a valid day of the week`);
          }

      }
    }();

    return await getJSON({
      hostname: "api.scratch.mit.edu",
      path: `/conference/schedule/${d}`
    });
  }

  async detailsFor(id) {
    return await getJSON({
      hostname: "api.scratch.mit.edu",
      path: `/conference/${id}/details`
    });
  }

}

class Users {
  async get(username) {
    return await getJSON({
      hostname: "api.scratch.mit.edu",
      path: `/users/${username}`
    });
  }

  async getFollowing(username) {
    let following = [];
    let offset = 0;

    while (true) {
      let batch = await getJSON({
        hostname: "api.scratch.mit.edu",
        path: `/users/${username}/following?limit=40&offset=${offset}`
      });
      following = following.concat(batch);

      if (batch.length < 40) {
        return following;
      }

      offset += 40;
    }
  }

  async getFollowers(username) {
    let followers = [];
    let offset = 0;

    while (true) {
      let batch = await getJSON({
        hostname: "api.scratch.mit.edu",
        path: `/users/${username}/followers?limit=40&offset=${offset}`
      });
      followers = followers.concat(batch);

      if (batch.length < 40) {
        return followers;
      }

      offset += 40;
    }
  }

  
    async getUserMessageCount(username) {
        let oData = await getJSON({
            hostname: "api.scratch.mit.edu",
            path: `/users/${username}/messages/count`
        });

        let data = JSON.parse(oData);

        return data.count;
    }   
}

class Site {
    async getHealthInfo() {
        return await getJSON({
            hostname: "api.scratch.mit.edu",
            path: "/health"
        });
    }

    async getNews() { // to be updated!
        return await getJSON({
            hostname: "api.scratch.mit.edu",
            path: "/news"
        });
    }
}

class Proxy {
    async getAllFeaturedData() {
        return await getJSON({
            hostname: "api.scratch.mit.edu",
            path: "/proxy/featured"
        });
    }

    async getFeatured() {
        let allData = await getJSON({
            hostname: "api.scratch.mit.edu",
            path: "/proxy/featured"
        });

        return allData.community_featured_projects;
    }

    async getNewest() {
        let allData = await getJSON({
            hostname: "api.scratch.mit.edu",
            path: "/proxy/featured"
        });

        return allData.community_newest_projects;
    }

    async getMostRemixed() {
        let allData = await getJSON({
            hostname: "api.scratch.mit.edu",
            path: "/proxy/featured"
        });

        return allData.community_most_remixed_projects;
    }

    async getScratchDesignStudio() {
        let allData = await getJSON({
            hostname: "api.scratch.mit.edu",
            path: "/proxy/featured"
        });

        return allData.scratch_design_studio;
    }

    async getCuratorTopProjects() {
        let allData = await getJSON({
            hostname: "api.scratch.mit.edu",
            path: "/proxy/featured"
        });

        return allData.curator_top_projects;
    }

    async getFeaturedStudios() {
        let allData = await getJSON({
            hostname: "api.scratch.mit.edu",
            path: "/proxy/featured"
        });

        return allData.community_featured_studios;
    }

    async getMostLoved() {
        let allData = await getJSON({
            hostname: "api.scratch.mit.edu",
            path: "/proxy/featured"
        });

        return allData.community_most_loved_projects;
    }

    // Finally done with the project stuff

    async getUserActivity(username) {
        return await getJSON({
            hostname: "api.scratch.mit.edu",
            path: `/proxy/users/${username}/activity`
        });
    }

    /*async getUserMessageCount(username) {                    No Longer Works
        let oData = await getJSON({
            hostname: "api.scratch.mit.edu",
            path: `/proxy/users/${username}/activity/count`
        });

        let data = JSON.parse(oData);

        return data.msg_count;
    }*/ 
}

module.exports = new API();
